Try to build Bazel (4.0) in a rootless environment on Ubuntu Jammy

Part of #1 

After the works in #2 we probably require Bazel 4.0 to build Drake on Ubuntu Jammy. The Bazel from source compilation uses a previous version of Bazel to bootstrap new Bazel so there is a chance that we can build Bazel `4.0` under a non root account using the system `3.5.1` version on Ubuntu Jammy and use that version to compile Drake instead of the system one.

A quick run for testing shown that it should be possible but there are compilers errors:
```sh
#10 41.75 INFO: Analyzed target //src:bazel-dev (326 packages loaded, 8662 targets configured).
#10 41.76 INFO: Found 1 target...
#10 41.91 [0 / 638] [Prepa] Writing file src/embedded_tools_jdk_allmodules.params
#10 49.06 [19 / 834] Compiling external/com_google_protobuf/src/google/protobuf/arena.cc; 1s processwrapper-sandbox ... (3 actions, 2 running)
#10 56.50 [37 / 834] Compiling external/com_google_protobuf/src/google/protobuf/generated_message_table_driven_lite.cc; 2s processwrapper-sandbox ... (4 actions, 3 running)
#10 64.69 INFO: From Compiling external/com_google_protobuf/src/google/protobuf/message_lite.cc [for host]:
#10 64.71 In file included from /usr/include/string.h:535,
#10 64.71                  from external/com_google_protobuf/src/google/protobuf/stubs/port.h:39,
#10 64.71                  from external/com_google_protobuf/src/google/protobuf/stubs/macros.h:34,
#10 64.71                  from external/com_google_protobuf/src/google/protobuf/stubs/common.h:46,
#10 64.71                  from external/com_google_protobuf/src/google/protobuf/message_lite.h:45,
#10 64.71                  from external/com_google_protobuf/src/google/protobuf/message_lite.cc:36:
#10 64.71 In function 'void* memcpy(void*, const void*, size_t)',
#10 64.71     inlined from 'google::protobuf::uint8* google::protobuf::io::EpsCopyOutputStream::WriteRaw(const void*, int, google::protobuf::uint8*)' at external/com_google_protobuf/src/google/protobuf/io/coded_stream.h:699:16,
#10 64.71     inlined from 'virtual google::protobuf::uint8* google::protobuf::internal::ImplicitWeakMessage::_InternalSerialize(google::protobuf::uint8*, google::protobuf::io::EpsCopyOutputStream*) const' at external/com_google_protobuf/src/google/protobuf/implicit_weak_message.h:85:28,
#10 64.71     inlined from 'bool google::protobuf::MessageLite::SerializePartialToZeroCopyStream(google::protobuf::io::ZeroCopyOutputStream*) const' at external/com_google_protobuf/src/google/protobuf/message_lite.cc:419:30:
#10 64.71 /usr/include/x86_64-linux-gnu/bits/string_fortified.h:29:33: warning: 'void* __builtin___memcpy_chk(void*, const void*, long unsigned int, long unsigned int)' specified size between 18446744071562067968 and 18446744073709551615 exceeds maximum object size 9223372036854775807 [-Wstringop-overflow=]
#10 64.71    29 |   return __builtin___memcpy_chk (__dest, __src, __len,
#10 64.71       |          ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
#10 64.71    30 |                                  __glibc_objsize0 (__dest));
#10 64.71       |                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~
#10 64.73 [59 / 834] Compiling external/com_google_protobuf/src/google/protobuf/repeated_field.cc [for host]; 2s processwrapper-sandbox ... (4 actions, 3 running)
#10 74.03 [76 / 834] Compiling external/com_google_protobuf/src/google/protobuf/generated_message_util.cc [for host]; 3s processwrapper-sandbox ... (4 actions, 3 running)
#10 84.76 [171 / 834] Compiling external/com_github_grpc_grpc/src/core/lib/debug/trace.cc; 0s processwrapper-sandbox ... (4 actions, 3 running)
#10 91.23 ERROR: /home/buildfarm/.cache/bazel/_bazel_buildfarm/ee8532c5c69dfff0fd0d0cf9658f23d3/external/com_google_absl/absl/synchronization/BUILD.bazel:30:11: C++ compilation of rule '@com_google_absl//absl/synchronization:graphcycles_internal' failed (Exit 1) gcc failed: error executing command /usr/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer '-std=c++0x' -MD -MF ... (remaining 31 argument(s) skipped)
#10 91.23 
#10 91.23 Use --sandbox_debug to see verbose messages from the sandbox
#10 91.23 external/com_google_absl/absl/synchronization/internal/graphcycles.cc: In member function 'void absl::lts_2020_02_25::synchronization_internal::GraphCycles::RemoveNode(void*)':
#10 91.23 external/com_google_absl/absl/synchronization/internal/graphcycles.cc:451:26: error: 'numeric_limits' is not a member of 'std'
#10 91.23   451 |   if (x->version == std::numeric_limits<uint32_t>::max()) {
#10 91.23       |                          ^~~~~~~~~~~~~~
#10 91.23 external/com_google_absl/absl/synchronization/internal/graphcycles.cc:451:49: error: expected primary-expression before '>' token
#10 91.23   451 |   if (x->version == std::numeric_limits<uint32_t>::max()) {
#10 91.23       |                                                 ^
#10 91.23 external/com_google_absl/absl/synchronization/internal/graphcycles.cc:451:52: error: '::max' has not been declared; did you mean 'std::max'?
#10 91.23   451 |   if (x->version == std::numeric_limits<uint32_t>::max()) {
#10 91.23       |                                                    ^~~
#10 91.23       |                                                    std::max
#10 91.23 In file included from /usr/include/c++/11/algorithm:62,
#10 91.23                  from external/com_google_absl/absl/synchronization/internal/graphcycles.cc:38:
#10 91.23 /usr/include/c++/11/bits/stl_algo.h:3467:5: note: 'std::max' declared here
#10 91.23  3467 |     max(initializer_list<_Tp> __l, _Compare __comp)
#10 91.23       |     ^~~
#10 91.37 Target //src:bazel-dev failed to build
#10 91.38 Use --verbose_failures to see the command lines of failed build steps.
```

Seems to be reported in https://github.com/bazelbuild/bazel/issues/12756. Still open.

--- a/experiments/bootstrap_bazel4_on_jammy/Dockerfile
+++ b/experiments/bootstrap_bazel4_on_jammy/Dockerfile
@@ -20,7 +20,7 @@ RUN useradd -u ${uid} -g ${group} -s /bin/sh -m ${user}
 USER ${uid}:${gid}
 WORKDIR /home/buildfarm
 
-RUN wget https://github.com/bazelbuild/bazel/archive/refs/tags/4.2.3.tar.gz \
+RUN wget -q https://github.com/bazelbuild/bazel/archive/refs/tags/4.2.3.tar.gz \
     && tar xzf 4.2.3.tar.gz \
     && echo "a9ef01b744d2fdcd688dcc9ecc25ad6dffaf4a9967058303fd27a5b41670b7d7 4.2.3.tar.gz" | sha256sum -c
 WORKDIR bazel-4.2.3

--- a/experiments/bootstrap_bazel4_on_jammy/Dockerfile
+++ b/experiments/bootstrap_bazel4_on_jammy/Dockerfile
@@ -25,7 +25,7 @@ RUN wget https://github.com/bazelbuild/bazel/archive/refs/tags/4.2.3.tar.gz \
     && echo "a9ef01b744d2fdcd688dcc9ecc25ad6dffaf4a9967058303fd27a5b41670b7d7 4.2.3.tar.gz" | sha256sum -c
 WORKDIR bazel-4.2.3
 COPY patches/absl_ghrp.patch absl_ghrp.patch
-COPY pacthes/ijar.patch ijar.patch
+COPY patches/ijar.patch ijar.patch
 RUN patch -p1 < absl_ghrp.patch \
     && patch -p1 < ijar.patch
 RUN bazel build //src:bazel-dev

--- a/experiments/bootstrap_bazel4_on_jammy/Dockerfile
+++ b/experiments/bootstrap_bazel4_on_jammy/Dockerfile
@@ -1,0 +1,31 @@
+FROM ubuntu:jammy
+MAINTAINER Jose Luis Rivero <jrivero@openrobotics.org>
+ENV LANG C
+ENV LC_ALL C
+ARG DEBIAN_FRONTEND=noninteractive
+RUN apt-get update && apt-get install -y \
+ bazel-bootstrap \
+ wget \
+ unzip \
+ zip \
+ zlib1g-dev \
+ && rm -rf /var/lib/apt/lists/*
+
+ARG user=buildfarm
+ARG group=buildfarm
+ARG uid=1000
+ARG gid=1000
+RUN groupadd -g ${gid} ${group}
+RUN useradd -u ${uid} -g ${group} -s /bin/sh -m ${user}
+USER ${uid}:${gid}
+WORKDIR /home/buildfarm
+
+RUN wget https://github.com/bazelbuild/bazel/archive/refs/tags/4.2.3.tar.gz \
+    && tar xzf 4.2.3.tar.gz \
+    && echo "a9ef01b744d2fdcd688dcc9ecc25ad6dffaf4a9967058303fd27a5b41670b7d7 4.2.3.tar.gz" | sha256sum -c
+WORKDIR bazel-4.2.3
+COPY patches/absl_ghrp.patch absl_ghrp.patch
+COPY pacthes/ijar.patch ijar.patch
+RUN patch -p1 < absl_ghrp.patch \
+    && patch -p1 < ijar.patch
+RUN bazel build //src:bazel-dev

--- a/experiments/bootstrap_bazel4_on_jammy/patches/absl_ghrp.patch
+++ b/experiments/bootstrap_bazel4_on_jammy/patches/absl_ghrp.patch
@@ -1,0 +1,49 @@
+Description: Add missing includes to abseil-cpp in order to build with gcc11
+Origin: https://gitlab.alpinelinux.org/alpine/aports/-/blob/2efd9732b4cd406ea424e5af71c915092a801e8c/testing/bazel4/patch_ftbfs_gcc11_4.patch
+Acked-by: Jose Luis Rivero <jrivero@openrobotics.org>
+
+--- a/third_party/grpc/grpc_1.33.1.patch
++++ b/third_party/grpc/grpc_1.33.1.patch
+@@ -58,6 +58,14 @@ index 09fcad95a2..9b737e5deb 100644
+      )
+  
+      native.bind(
++@@ -245,6 +245,7 @@ def grpc_deps():
++                 "https://storage.googleapis.com/grpc-bazel-mirror/github.com/abseil/abseil-cpp/archive/df3ea785d8c30a9503321a3d35ee7d35808f190d.tar.gz",
++                 "https://github.com/abseil/abseil-cpp/archive/df3ea785d8c30a9503321a3d35ee7d35808f190d.tar.gz",
++             ],
+++            patches = ["@com_github_grpc_grpc//:third_party/abseil-cpp/absl.patch"],
++         )
++
++     if "bazel_toolchains" not in native.existing_rules():
+ diff --git a/bazel/grpc_extra_deps.bzl b/bazel/grpc_extra_deps.bzl
+ index 4c1dfad2e8..f63c54ddef 100644
+ --- a/bazel/grpc_extra_deps.bzl
+@@ -120,3 +128,25 @@ index c047f0c515..7c24fbc617 100644
+          ":windows": "@com_github_grpc_grpc//third_party/cares:config_windows/ares_config.h",
+          ":android": "@com_github_grpc_grpc//third_party/cares:config_android/ares_config.h",
+          "//conditions:default": "@com_github_grpc_grpc//third_party/cares:config_linux/ares_config.h",
++
++--- /dev/null
+++++ b/third_party/abseil-cpp/absl.patch
++@@ -0,0 +1,18 @@
+++0e2c62da1dcaf6529abab952bdcc96c6de2d9506 by Abseil Team <absl-team@google.com>:
+++
+++Add missing <limits> include
+++
+++PiperOrigin-RevId: 339054753
+++
+++--
+++
+++--- absl/synchronization/internal/graphcycles.cc
++++++ absl/synchronization/internal/graphcycles.cc
+++@@ -37,6 +37,7 @@
+++ 
+++ #include <algorithm>
+++ #include <array>
++++#include <limits>
+++ #include "absl/base/internal/hide_ptr.h"
+++ #include "absl/base/internal/raw_logging.h"
+++ #include "absl/base/internal/spinlock.h"
+
+

--- a/experiments/bootstrap_bazel4_on_jammy/patches/ijar.patch
+++ b/experiments/bootstrap_bazel4_on_jammy/patches/ijar.patch
@@ -1,0 +1,36 @@
+Description: Add missing includes to ijar in order to support gcc11 build
+Origin: https://gitlab.alpinelinux.org/alpine/aports/-/blob/2efd9732b4cd406ea424e5af71c915092a801e8c/testing/bazel4/patch_ftbfs_gcc11_1.patch
+Acked-by: Jose Luis Rivero <jrivero@openrobotics.org>
+
+diff --git a/third_party/ijar/mapped_file_unix.cc b/third_party/ijar/mapped_file_unix.cc
+index 6e3a908..65179e3 100644
+--- a/third_party/ijar/mapped_file_unix.cc
++++ b/third_party/ijar/mapped_file_unix.cc
+@@ -15,10 +15,11 @@
+ #include <errno.h>
+ #include <fcntl.h>
+ #include <stdio.h>
+-#include <unistd.h>
+ #include <sys/mman.h>
++#include <unistd.h>
+ 
+ #include <algorithm>
++#include <limits>
+ 
+ #include "third_party/ijar/mapped_file.h"
+ 
+diff --git a/third_party/ijar/zlib_client.h b/third_party/ijar/zlib_client.h
+index ed66163..0a917ff 100644
+--- a/third_party/ijar/zlib_client.h
++++ b/third_party/ijar/zlib_client.h
+@@ -17,6 +17,9 @@
+ 
+ #include <limits.h>
+ 
++#include <limits>
++#include <stdexcept>
++
+ #include "third_party/ijar/common.h"
+ 
+ namespace devtools_ijar {
+(END)

